python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -220,15 +220,153 @@ predicate jumpStep(Node pred, Node succ) {
 // Field flow
 //--------
 /**
- * Holds if data can flow from `node1` to `node2` via an assignment to
+ * Holds if data can flow from `nodeFrom` to `nodeTo` via an assignment to
  * content `c`.
  */
-predicate storeStep(Node node1, Content c, Node node2) { none() }
+predicate storeStep(Node nodeFrom, Content c, Node nodeTo) {
+  // Sequence
+  //   `(..., 42, ...)`
+  //   or
+  //   `[..., 42, ...]`
+  //   nodeFrom is `42`, cfg node
+  //   nodeTo is the sequence, say `[..., 42, ...]`, cfg node
+  //   c denotes list or c denotes tuple and index of 42
+  //
+  // List
+  nodeTo.(CfgNode).getNode().(ListNode).getAnElement() = nodeFrom.(CfgNode).getNode() and
+  c instanceof ListElementContent
+  or
+  // Tuple
+  exists(int n |
+    nodeTo.(CfgNode).getNode().(TupleNode).getElement(n) = nodeFrom.(CfgNode).getNode() and
+    c.(TupleElementContent).getIndex() = n
+  )
+  or
+  // Dict
+  exists(KeyValuePair item |
+    item = nodeTo.(CfgNode).getNode().(DictNode).getNode().(Dict).getAnItem() and
+    nodeFrom.(CfgNode).getNode().getNode() = item.getValue() and
+    c.(DictionaryElementContent).getKey() = item.getKey().(StrConst).getS()
+  )
+  or
+  //
+  // Comprehension
+  //   `[x+1 for x in l]`
+  //   nodeFrom is `x+1`, cfg node
+  //   nodeTo is `[x+1 for x in l]`, cfg node
+  //   c denotes list or set or dictionary without index
+  //
+  // List
+  nodeTo.(CfgNode).getNode().getNode().(ListComp).getElt() = nodeFrom.(CfgNode).getNode().getNode() and
+  c instanceof ListElementContent
+  or
+  // Set
+  nodeTo.(CfgNode).getNode().getNode().(SetComp).getElt() = nodeFrom.(CfgNode).getNode().getNode() and
+  c instanceof SetElementContent
+  or
+  // Dictionary
+  nodeTo.(CfgNode).getNode().getNode().(DictComp).getElt() = nodeFrom.(CfgNode).getNode().getNode() and
+  c instanceof DictionaryElementAnyContent
```

Comment:
An alternative here is to introduce synthetic nodes that record the key that was read, so you get more precision. So something like this will work:
```
dict["a"] = "taint"
dict["b"] = "no taint"
dict = {k:v+"" for (k,v) in dict.items()}
Sink(dict["b"]) // not tainted
```

You will then add a `readStep` from `dict` with content `TDictionaryElementContent(key)` to `TSynthesizedDictCompNode(key, v)`, a local step from `TSynthesizedDictCompNode(key, v)` to `TSynthesizedDictCompNode(key, v+"")` and a `storeStep` with content `TDictionaryElementContent(key)` from `TSynthesizedDictCompNode(key, v+"")` to the entire comprehension.

Reply:
This is an interesting use of synthetic nodes that we should keep in mind. It sparked thoughts that I wrote down [here](https://github.com/github/codeql-python-team/issues/134). It is probably too much precision to strive for right now, but perhaps we should add some test cases to record the ambition :-)